src/main/java/bio/terra/service/snapshot/flight/export/SnapshotExportValidateStep.java

```diff
@@ -0,0 +1,110 @@
+    Dataset dataset = snapshot.getFirstSnapshotSource().getDataset();
+    List<SnapshotTable> tables = snapshot.getTables();
+    List<String> tablesWithoutPrimaryKeys = new ArrayList<>();
+    List<String> tablesWithDuplicates =
+        tables.stream()
+            .map(
+                table -> {
+                  String tableName = table.getName();
+                  Optional<DatasetTable> datasetTable = dataset.getTableByName(tableName);
+                  if (datasetTable.isEmpty()) {
+                    throw new CorruptMetadataException(
+                        String.format(
+                            "Could not find '%s' table in the snapshot's dataset", tableName));
+                  }
+                  DatasetTable realDatasetTable = datasetTable.get();
+                  List<Column> primaryKeyColumns = realDatasetTable.getPrimaryKey();
+                  if (primaryKeyColumns.isEmpty()) {
+                    tablesWithoutPrimaryKeys.add(table.getName());
+                    return null;
+                  }
+
+                  boolean hasDuplicates;
+                  try {
+                    hasDuplicates =
+                        bigQueryPdao.hasDuplicatePrimaryKeys(
+                            snapshot, primaryKeyColumns, tableName);
+                  } catch (InterruptedException e) {
+                    throw new GoogleResourceException(
+                        String.format(
+                            "Failed to query BigQuery for duplicate rows in snapshot table '%s'",
+                            tableName));
+                  }
+                  if (hasDuplicates) {
+                    return table.getName();
+                  } else {
+                    return null;
+                  }
+                })
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+
+    if (tablesWithDuplicates.isEmpty() && tablesWithoutPrimaryKeys.isEmpty()) {
+      return StepResult.getStepResultSuccess();
+    } else {
+      StringBuilder sb = new StringBuilder();
+      sb.append("Validating snapshot for export failed. ");
+      if (!tablesWithDuplicates.isEmpty()) {
+        int numTables = tablesWithDuplicates.size();
+        String tableNames = String.join(", ", tablesWithDuplicates);
+        sb.append(
+            String.format(
+                "%d table(s) had rows with duplicate primary keys [%s]. ", numTables, tableNames));
+        sb.append(
+            "To export, please create a new snapshot, ensuring there are rows with duplicate primary keys. ");
+      }
+      if (!tablesWithoutPrimaryKeys.isEmpty()) {
+        int numTables = tablesWithoutPrimaryKeys.size();
+        String tableNames = String.join(", ", tablesWithoutPrimaryKeys);
+        sb.append(String.format("%d table(s) had no primary keys [%s]. ", numTables, tableNames));
+        sb.append("To export, please ensure every table being exported has a primary key. ");
```

Comment:
I don't think that we want this behavior.  I think it's reasonable for a snapshot to have some tables with PKs and some without.  I don't think we want to fail in this case